runtime/src/account_rent_state.rs

```diff
@@ -0,0 +1,165 @@
+use {
+    crate::bank::TransactionAccountRefCell,
+    enum_iterator::IntoEnumIterator,
+    solana_sdk::{account::ReadableAccount, native_loader, rent::Rent, sysvar},
+};
+
+#[derive(Debug, PartialEq, IntoEnumIterator)]
+pub(crate) enum RentState {
+    Uninitialized, // account.lamports == 0
+    NativeOrSysvar,
+    EmptyData,  // account.data.len == 0
+    RentPaying, // 0 < account.lamports < rent-exempt-minimum for non-zero data size
+    RentExempt, // account.lamports >= rent-exempt-minimum for non-zero data size
+}
+
+impl RentState {
+    pub(crate) fn from_account(account: &TransactionAccountRefCell, rent: &Rent) -> Self {
+        let account = account.1.borrow();
+        if account.lamports() == 0 {
+            Self::Uninitialized
+        } else if native_loader::check_id(account.owner()) || sysvar::is_sysvar_id(account.owner())
+        {
+            Self::NativeOrSysvar
+        } else if account.data().is_empty() {
+            Self::EmptyData
+        } else if !rent.is_exempt(account.lamports(), account.data().len()) {
+            Self::RentPaying
+        } else {
+            Self::RentExempt
+        }
+    }
+
+    pub(crate) fn transition_allowed_from(&self, pre_rent_state: &RentState) -> bool {
+        // Only a legacy RentPaying account may end in the RentPaying state after message processing
+        !(self == &Self::RentPaying && pre_rent_state != &Self::RentPaying)
```

Comment:
Actually, this is problematic if you started out with an `EmptyData` account. 

For instance, this wouldn't work with system_instruction::allocate